lib/runtime-core/src/alternative_stack.rs

```diff
@@ -0,0 +1,462 @@
+        ctx.ret.take().unwrap()
+    }
+}
+
+extern "C" fn signal_trap_handler(
+    signum: ::nix::libc::c_int,
+    siginfo: *mut siginfo_t,
+    ucontext: *mut c_void,
+) {
+    unsafe {
+        let fault = get_fault_info(siginfo as _, ucontext);
+
+        let mut unwind_result: Box<dyn Any> = Box::new(());
+
+        let should_unwind = allocate_and_run(1048576, || {
+            let mut is_suspend_signal = false;
+
+            match Signal::from_c_int(signum) {
+                Ok(SIGTRAP) => {
+                    // breakpoint
+                    let out: Option<Result<(), Box<dyn Any>>> = with_breakpoint_map(|bkpt_map| {
+                        bkpt_map.and_then(|x| x.get(&(fault.ip as usize))).map(|x| {
+                            x(BkptInfo {
+                                fault: Some(&fault),
+                            })
+                        })
+                    });
+                    match out {
+                        Some(Ok(())) => {
+                            return false;
+                        }
+                        Some(Err(e)) => {
+                            unwind_result = e;
+                            return true;
+                        }
+                        None => {}
+                    }
+                }
+                Ok(SIGSEGV) | Ok(SIGBUS) => {
+                    if fault.faulting_addr as usize == get_wasm_interrupt_signal_mem() as usize {
+                        is_suspend_signal = true;
+                        clear_wasm_interrupt();
+                        INTERRUPT_SIGNAL_DELIVERED.store(false, Ordering::SeqCst);
+                    }
+                }
+                _ => {}
+            }
+
+            // TODO: make this safer
+            let ctx = &mut *(fault.known_registers[X64Register::GPR(GPR::R15).to_index().0].unwrap()
+                as *mut vm::Ctx);
+            let rsp = fault.known_registers[X64Register::GPR(GPR::RSP).to_index().0].unwrap();
+
+            let msm = (*ctx.module)
+                .runnable_module
+                .get_module_state_map()
+                .unwrap();
+            let code_base = (*ctx.module).runnable_module.get_code().unwrap().as_ptr() as usize;
+            let es_image = read_stack(
+                &msm,
+                code_base,
+                rsp as usize as *const u64,
+                fault.known_registers,
+                Some(fault.ip as usize as u64),
+            );
+
+            if is_suspend_signal {
+                let image = build_instance_image(ctx, es_image);
+                unwind_result = Box::new(image);
+            } else {
+                use colored::*;
+                if es_image.frames.len() > 0 {
+                    eprintln!(
+                        "\n{}",
+                        "Wasmer encountered an error while running your WebAssembly program."
+                            .bold()
+                            .red()
+                    );
```

Comment:
Should we avoid using stderr and prefer providing users with an `Err` value to have them decide if stderr should be used?

Reply:
This is designed to be similar to how the Rust panic backtrace works. Some more discussion and thinking on this is needed though and we can change this in the future.